File: app/src/main/java/com/tokyonth/installer/adapter/SettingsAdapter.java
package com.tokyonth.installer.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.kyleduo.switchbutton.SwitchButton;
import com.tokyonth.installer.Constants;
import com.tokyonth.installer.R;
import com.tokyonth.installer.bean.SettingsBean;
import com.tokyonth.installer.utils.SPUtils;
import com.tokyonth.installer.widget.BurnRoundView;

import java.util.ArrayList;

public class SettingsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private ArrayList<SettingsBean> list;
    private OnItemSwitchClick onItemSwitchClick;

    public SettingsAdapter(ArrayList<SettingsBean> list) {
        this.list = list;
    }

    public void setOnItemClick(OnItemSwitchClick onItemSwitchClick) {
        this.onItemSwitchClick = onItemSwitchClick;
    }

    public interface OnItemSwitchClick {
        void onItemClick(View view, int pos, boolean bool);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.rv_settings_item, parent, false);
        return new CommonViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, final int position) {
        if (holder instanceof CommonViewHolder) {
            SettingsBean bean = list.get(position);
            ((CommonViewHolder) holder).title.setText(bean.getTitle());
            ((CommonViewHolder) holder).sub.setText(bean.getSub());
            ((CommonViewHolder) holder).icon.setBurnSrc(bean.getIcon(), bean.getColor());
            ((CommonViewHolder) holder).switchBtn.setOnCheckedChangeListener((compoundButton, isChecked) -> onItemSwitchClick.onItemClick(compoundButton, position, isChecked));
            switch (position) {
                case 0:
                    ((CommonViewHolder) holder).switchBtn.setChecked((boolean)SPUtils.getData(Constants.SP_SHOW_PERM, true));
                    break;
                case 1:
                    ((CommonViewHolder) holder).switchBtn.setChecked((boolean)SPUtils.getData(Constants.SP_SHOW_ACT, true));
                    break;
                case 2:
                    ((CommonViewHolder) holder).switchBtn.setChecked((boolean)SPUtils.getData(Constants.SP_VIBRATE, false));
                    break;
            }
        }
    }

    @Override
    public int getItemCount() {
        return list == null ? 0 : list.size();
    }

    @Override
    public long getItemId(int position) {
        return super.getItemId(position);
    }

    static class CommonViewHolder extends RecyclerView.ViewHolder {

        private TextView title;
        private TextView sub;
        private BurnRoundView icon;
        private SwitchButton switchBtn;

        CommonViewHolder(View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.settings_item_title);
            sub = itemView.findViewById(R.id.settings_item_sub);
            icon = itemView.findViewById(R.id.settings_item_icon);
            switchBtn = itemView.findViewById(R.id.settings_item_switch);
        }
    }

}

File: app/src/main/java/com/tokyonth/installer/adapter/SettingsAdapter.kt
package com.tokyonth.installer.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import androidx.recyclerview.widget.RecyclerView

import com.tokyonth.installer.R
import com.tokyonth.installer.data.SPDataManager
import com.tokyonth.installer.data.SettingsEntity
import com.tokyonth.installer.databinding.ItemSettingPerfBinding
import com.tokyonth.installer.utils.ktx.click
import com.tokyonth.installer.utils.ktx.string

import java.util.ArrayList

class SettingsAdapter : RecyclerView.Adapter<SettingsAdapter.SettingsViewHolder>() {

    private val list: MutableList<SettingsEntity> = ArrayList()

    init {
        list.apply {
            add(
                SettingsEntity(
                    string(R.string.title_show_perm),
                    string(R.string.summary_show_perm),
                    R.drawable.round_security_24,
                    SPDataManager.instance.isShowPermission()
                )
            )
            add(
                SettingsEntity(
                    string(R.string.title_show_activity),
                    string(R.string.summary_show_activity),
                    R.drawable.round_layers_24,
                    SPDataManager.instance.isShowActivity()
                )
            )
            add(
                SettingsEntity(
                    string(R.string.default_silent),
                    string(R.string.default_silent_sub),
                    R.drawable.round_silence_24,
                    SPDataManager.instance.isDefaultSilent()
                )
            )
            add(
                SettingsEntity(
                    string(R.string.auto_del_apk_title),
                    string(R.string.auto_delete_apk),
                    R.drawable.round_delete_forever_24,
                    SPDataManager.instance.isAutoDel()
                )
            )
            add(
                SettingsEntity(
                    string(R.string.follow_system_night_mode),
                    string(R.string.follow_system_night_mode_sub),
                    R.drawable.round_auto_mode_24,
                    SPDataManager.instance.isFollowSystem()
                )
            )
            add(
                SettingsEntity(
                    string(R.string.install_mode),
                    SPDataManager.instance.getInstallName(),
                    R.drawable.round_play_for_work_24,
                    false
                )
            )
        }
    }

    interface OnItemActionListener {

        fun onSwitch(view: CompoundButton, position: Int, bool: Boolean)

        fun onClick(position: Int)

    }

    private var onItemClickListener: OnItemActionListener? = null

    fun setOnItemActionListener(onItemActionListener: OnItemActionListener) {
        this.onItemClickListener = onItemActionListener
    }

    fun updateInstallMode() {
        list[list.size - 1].sub = SPDataManager.instance.getInstallName()
        notifyItemChanged(list.size - 1)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SettingsViewHolder {
        val binding =
            ItemSettingPerfBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return SettingsViewHolder(binding)
    }

    override fun onBindViewHolder(holder: SettingsViewHolder, position: Int) {
        holder.bind(list[position], onItemClickListener)
    }

    override fun getItemCount(): Int {
        return list.size
    }

    inner class SettingsViewHolder(private val binding: ItemSettingPerfBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(
            data: SettingsEntity,
            action: OnItemActionListener?
        ) {
            binding.itemSettingTitle.text = data.title
            binding.itemSettingSub.text = data.sub
            binding.itemSettingIcon.setImageResource(data.icon)

            binding.itemSettingSwitch.run {
                isChecked = data.selected
                if (bindingAdapterPosition == 5) {
                    visibility = View.GONE
                }

                setOnCheckedChangeListener { view, isChecked ->
                    action?.onSwitch(view, bindingAdapterPosition, isChecked)
                }
            }

            binding.root.click {
                action?.onClick(bindingAdapterPosition)
            }
        }
    }

}
